### PR TITLE
preprocess: utils to deal with BOM, DOS, UTF-16

### PR DESCRIPTION
**Pull request by ahmetalpbalkan**
Contains the utilities to detect & sanitize text files (i.e. scripts).

- bom.go: removal of bom chars, utf16 to utf8 conversion
- dos.go: removal of DOS line endings
- file.go: detection of script files via file ext or shebang

and bunch of test files (inherited from old repo)

### CODE REVIEW

File: preprocess/bom.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+)
+
+var bomSequences = [][]byte{
+	{'\xef', '\xbb', '\xbf'}, // python: codecs.BOM_UTF8
+	{'\xff', '\xfe'},         // python: codecs.BOM, codecs.BOM_LE, codecs.BOM_UTF16_LE
+	{'\xfe', '\xff'},         // python: codecs.BOM_BE, codecs.BOM_UTF16_BE
+}
+
+var (
+	utf16le = []byte{'\xff', '\xfe'}
```
Comment (boumenot) [inline preprocess/bom.go]: utf8 too?

Comment (ahmetalpbalkan) [inline preprocess/bom.go]: ah I'm realizing these are leftover..

File: preprocess/file.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	"unicode"
+
+	"github.com/pkg/errors"
+)
+
+const peekLen = 64 // look at first N bytes to figure out if it has shebang
+
+// textExtensions is predefined list of script and text
+// file extensions.
+var textExtensions = []string{
```
Comment (boumenot) [inline preprocess/file.go]: is this from the current version?  .rb for ruby?

File: preprocess/file_test.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	require.True(t, hasShebang([]byte("#! foo")), "shebang")
+
+	script := "\r\n  \t#!foo"
+	require.True(t, hasShebang([]byte(script)), "whitespace")
+
+	noscript := "    !$"
+	require.False(t, hasShebang([]byte(noscript)), "not shebang")
+
+	bomdata := append(bomSequences[1], []byte("    \t \n#!booyah")...)
+	require.True(t, hasShebang(bomdata), true)
```
Comment (boumenot) [inline preprocess/file_test.go]: I would have said false.  Shebang should be at the absolute beginning, minus the BOM.  (Did I interpret that correctly?)

**Comment by boumenot** [comment]
Few questions, but LGTM.

File: preprocess/file.go (hunk 2 above)
Comment (ahmetalpbalkan) [inline preprocess/file.go]: they are mostly from the current version. I think I just added .txt and .pl. Can be extended but I'd say let's not if there's no clear need.

**Pull request merged**